### PR TITLE
Wip ab/llava next cp sp seq packing

### PR DESCRIPTION
**Pull request by abhinavg4**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/multimodal/data/energon/base.py [hunk 1]
```diff
@@ -297,7 +296,13 @@ def state_dict(self) -> Dict[str, Any]:
             dataloader_obj = self.trainer.train_dataloader
 
             state = []
-            if torch.distributed.get_rank() == parallel_state.get_model_parallel_src_rank():
+            # All ranks should be zero except the dp rank.
```
Comment (yashaswikarnati) [inline nemo/collections/multimodal/data/energon/base.py:299]: wonder if this change part of this PR?

File: nemo/collections/vlm/neva/data/sequence_packing.py [hunk 2]
```diff
@@ -76,6 +76,34 @@ def greedy_knapsack(item_sizes: List[int], samples: List, max_capacity: int) ->
     return knapsacks
 
 
+def predict_seq_len_llava_next(
```
Comment (yashaswikarnati) [inline nemo/collections/vlm/neva/data/sequence_packing.py:79]: may be we can move this to llava next

File: nemo/collections/vlm/neva/data/sequence_packing.py [hunk 3]
```diff
@@ -155,3 +190,72 @@ def convert_to_packed(
     )
 
     return packed_tokens, packed_labels, packed_position_ids, packed_loss_mask, packed_seq_params
+
+
+def convert_to_packed_llava_next(
```
Comment (yashaswikarnati) [inline nemo/collections/vlm/neva/data/sequence_packing.py:195]: can this be moved as well?

File: nemo/collections/multimodal/data/energon/base.py (hunk 1 above)
Comment (abhinavg4) [reply nemo/collections/multimodal/data/energon/base.py:299]: Probably not. It should be a seperate PR but since this PR is already up, I think we can let it be part of this.

File: nemo/collections/vlm/neva/data/sequence_packing.py [hunk 4]
```diff
@@ -93,6 +93,13 @@ def predict_seq_len(instance_tokens: torch.Tensor, num_image_embeddings_per_tile
     return seqlen
 
 
+def predict_seq_len_with_padding(instance_tokens: torch.Tensor, pad_to_multiple_of: int = 64) -> int:
```
Comment (yaoyu-33) [inline nemo/collections/vlm/neva/data/sequence_packing.py:96]: can we merge this with `predict_seq_len`? or it's for a different input use case?

File: nemo/collections/vlm/llava_next/model/base.py [hunk 5]
```diff
@@ -338,31 +326,46 @@ def forward(
...
+            attention_mask=None,
             decoder_input=combined_embeddings,
             labels=final_labels,
             inference_params=inference_params,
             runtime_gather_output=runtime_gather_output,
+            packed_seq_params=packed_seq_params,
         )
 
-        if labels is None or loss_mask is None:
+        if labels is None:
```
Comment (yaoyu-33) [inline nemo/collections/vlm/llava_next/model/base.py:368]: why is this change?

File: nemo/collections/vlm/llava_next/data/utils.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+# limitations under the License.
+
+from typing import List
+
+import torch
+import torch.nn.functional as F
+from megatron.core.packed_seq_params import PackedSeqParams
+
+
+def convert_to_packed_llava_next(
```
Comment (yaoyu-33) [inline nemo/collections/vlm/llava_next/data/utils.py:22]: is it possible to re-use the existing one? what's the major problem of re-using?

File: nemo/collections/vlm/llava_next/data/sample.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        used to determine which tokens or tiles are attended to during processing. Defaults to None.
+    '''
+
+    num_media_tiles: int = 0
+    image_sizes: torch.tensor = field(default_factory=lambda: torch.tensor([]))
+    attention_mask: Optional[torch.tensor] = None
+
+
+@dataclass
+class PackedLlavaNextTextSample(LlavaNextTextSample):
```
Comment (yashaswikarnati) [inline nemo/collections/vlm/llava_next/data/sample.py:44]: does this need to be a separate sample? wonder if we can merge with above

File: nemo/collections/vlm/llava_next/data/sample.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        attention_mask (Optional[torch.Tensor]): Attention mask. Defaults to None.
+    """
+
+    num_media_tiles: List[int] = field(default_factory=list)
+    image_sizes: torch.tensor = field(default_factory=lambda: torch.tensor([]))
+    attention_mask: Optional[torch.tensor] = None
+
+
+@dataclass
+class PackedLlavaNextTextRawBatch(LlavaNextTextRawBatch):
```
Comment (yashaswikarnati) [inline nemo/collections/vlm/llava_next/data/sample.py:72]: same comment as sample,

File: nemo/collections/vlm/llava_next/data/task_encoder.py [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+                    f"Set micro batch size to 1 (currently {len(samples)})\n"
+                    f"Set global batch size to `global_batch_size // {len(samples)}` "
+                    f"Set packed sequence length to `original_sample_seq_len * {len(samples)}` "
+                    f"(currently {self.packed_sequence_size}) \n"
+                    f"For details please visit "
+                    f"https://docs.nvidia.com/nemo-framework/user-guide/latest/sft_peft/packed_sequence.html"
+                )
+            # The batching are taken care by packing.
+            sample = samples[0]
+            return PackedLlavaNextTextRawBatch(
```
Comment (yashaswikarnati) [inline nemo/collections/vlm/llava_next/data/task_encoder.py:106]: if we merge sample types, this might become easier

File: nemo/collections/vlm/llava_next/model/base.py [hunk 10]
```diff
@@ -294,34 +306,10 @@ def forward(
...
-                    - input_ids_text.shape[1]
-                )
-                if padded_seq_len != 0:
-                    input_ids_text = torch.nn.functional.pad(input_ids_text, (0, padded_seq_len))
-                    if position_ids is not None:
-                        position_ids = torch.nn.functional.pad(position_ids, (0, padded_seq_len))
+
             language_embeddings = self.language_model.embedding(
-                input_ids=input_ids_text, position_ids=position_ids
+                input_ids=input_ids_text, position_ids=None
```
Comment (yashaswikarnati) [inline nemo/collections/vlm/llava_next/model/base.py:311]: should the position_ids be passed as arg? wonder if fixing this to None is needed.

File: nemo/collections/vlm/llava_next/model/base.py [hunk 11]
```diff
@@ -338,31 +326,46 @@ def forward(
...
+            raise ValueError(
+                f"Image features and image tokens do not match: tokens: {n_image_tokens}, features {n_image_features}"
             )
-        )
+        special_image_mask = (input_ids == media_token_index).unsqueeze(-1)
+        special_image_mask = special_image_mask.expand_as(language_embeddings).to(language_embeddings.device)
+        media_embeddings = media_embeddings.to(language_embeddings.device, language_embeddings.dtype)
+        combined_embeddings = language_embeddings.masked_scatter(special_image_mask, media_embeddings)
+
+        final_labels = labels
```
Comment (yashaswikarnati) [inline nemo/collections/vlm/llava_next/model/base.py:340]: do we need reassignment here ?

File: nemo/collections/vlm/neva/data/sequence_packing.py (hunk 4 above)
Comment (abhinavg4) [reply nemo/collections/vlm/neva/data/sequence_packing.py:96]: Hi yes, They are completely different `predict_seq_len` is used in Neva whereas `predict_seq_len_with_padding` in llava-next. We can merge them moving forward ig.

File: scripts/vlm/llava_next_pretrain.py [hunk 12]
```diff
@@ -43,7 +44,9 @@ def main(args):
     mbs = args.mbs
     max_steps = args.max_steps
 
+    # For Interleaved data, the decoder sequence length needs to be higher than VQA samples
     decoder_seq_length = 4096
+    # decoder_seq_length = 40960
```
[the remark this answers is not in the document]
Comment (abhinavg4) [reply scripts/vlm/llava_next_pretrain.py:49]: For interleaved the docder_seq_length required is much larger. So I kept this and added a comment just above this.

File: nemo/collections/vlm/llava_next/model/base.py (hunk 5 above)
Comment (abhinavg4) [reply nemo/collections/vlm/llava_next/model/base.py:368]: Reverted

File: nemo/collections/vlm/llava_next/data/utils.py (hunk 6 above)
Comment (abhinavg4) [reply nemo/collections/vlm/llava_next/data/utils.py:22]: It's mainly the predict seq length call difference and `final_padding_to` argument. I will work on merging this in the refactoring soon.

File: nemo/collections/vlm/llava_next/data/sample.py (hunk 7 above)
Comment (abhinavg4) [reply nemo/collections/vlm/llava_next/data/sample.py:44]: Packed samples have some extra parameters. We can merge them with above with packed params optional but I think we plan to handle this during refactoring anyways.

File: nemo/collections/vlm/llava_next/model/base.py (hunk 10 above)
Comment (abhinavg4) [reply nemo/collections/vlm/llava_next/model/base.py:311]: I can revert it but I think for Neva and Llava next we assume that we use RoPE (other causal Embedding), so we don't need to pass `position_ids` explicitly. Even if you pass, they get ignored inside embedding I think

File: nemo/collections/vlm/llava_next/model/base.py (hunk 11 above)
Comment (abhinavg4) [reply nemo/collections/vlm/llava_next/model/base.py:340]: There is no reassignment. We can remove this assignment and use labels directly but I think to make code consistent, this is good. We use final_loss in all VLM models.

File: nemo/collections/vlm/llava_next/model/base.py (hunk 10 above)
Comment (yashaswikarnati) [reply nemo/collections/vlm/llava_next/model/base.py:311]: but what happens if some wants to use an LLM that needs position embeddings? doesn't hurt to pass. you can add a comment saying that its ignored if LLM uses rope

File: nemo/collections/vlm/llava_next/model/base.py (hunk 11 above)
Comment (yashaswikarnati) [reply nemo/collections/vlm/llava_next/model/base.py:340]: I don't think using extra unnecessary variables is necessary here, in other VLM's they have final labels because it gets changed?

File: nemo/collections/vlm/llava_next/data/sample.py (hunk 7 above)
Comment (yashaswikarnati) [reply nemo/collections/vlm/llava_next/data/sample.py:44]: is the scope of refactoring planned? I think these two are independent,

File: nemo/collections/vlm/llava_next/model/base.py (hunk 11 above)
Comment (abhinavg4) [reply nemo/collections/vlm/llava_next/model/base.py:340]: So for us too we modify this variable in CP

File: nemo/collections/vlm/llava_next/model/base.py (hunk 10 above)
Comment (abhinavg4) [reply nemo/collections/vlm/llava_next/model/base.py:311]: Sure

File: nemo/collections/vlm/llava_next/data/sample.py (hunk 7 above)
Comment (abhinavg4) [reply nemo/collections/vlm/llava_next/data/sample.py:44]: Yup. This is the first thing in plan.

Comment (abhinavg4) [reply nemo/collections/vlm/llava_next/data/sample.py:44]: I think Yu can share the plan with you. He had put something together for llama4 and we the refined it a bit.

File: nemo/collections/vlm/llava_next/model/base.py (hunk 11 above)
Comment (yashaswikarnati) [reply nemo/collections/vlm/llava_next/model/base.py:341]: oh ok sorry missed this, its not visible in the diff

**Review by yaoyu-33 — APPROVED** [review]
